README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# OLStrajr

<!-- badges: start -->
<!-- badges: end -->

The [Maximizing the Yield of Small Samples in Prevention Research: A
Review of General Strategies and Best
Practices](https://link.springer.com/article/10.1007/s11121-014-0542-7)
paper by Hopkin, Hoyle, and Gottfredson (2015) suggests that when
working with small sample sizes, analysts should leverage plots to
explore within-group associations between predictor variables and the
outcome variable. These plots are intended to provide detailed insights
about the data rather than to be used for generalizing to the
population.

The paper cites the work of [Carrig, Wirth, and Curran
(2004)](https://www.tandfonline.com/doi/abs/10.1207/S15328007SEM1101_9)
who proposed an easy-to-use SAS macro for visualizing person-specific
growth trajectories with repeated measures data.

The objective of the OLStrajr package is to bring the functionality of
this SAS macro, OLStraj, to R. While the original link to the macro in
Carrig’s paper is no longer active, it can still be accessed through the
[Wayback Machine](https://archive.org/web/).

## Installation

You can install the released version of OLStrajr from CRAN with:

``` r
install.packages("OLStrajr")
```

You can install the development version of OLStrajr as follows:

``` r
# install.packages("devtools")
devtools::install_github("mightymetrika/OLStrajr")
```

## Ratio of Robin Males to Females

This serves as an exemplary demonstration of using OLStrajr. It features
an analysis of data representing the yearly ratio of male to female
robins, as documented in [Birds: incomplete counts—five-minute bird
counts Version
1.0](https://www.doc.govt.nz/documents/science-and-technical/inventory-monitoring/im-toolbox-birds-incomplete-five-min-counts.pdf).
This data, collected at Walker Creek and Knobs Flat from August 2005 to
August 2009, provides a compelling case study

``` r
# Load package
library(OLStrajr)

# Get robins data
data(robins)

# Print robins data
robins
#>           site aug_05 aug_06 aug_07 aug_08 aug_09
#> 1 Walker_Creek   1.67   1.50   1.80    1.6   1.38
#> 2   Knobs_Flat   1.64   1.76   2.33    3.0   2.40
```

### OLStraj

To obtain the OLS trajectories, use the following code:

``` r
robins_traj <- OLStraj(data = robins,
                      idvarname = "site",
                      predvarname = "Year",
                      outvarname = "Ratio",
                      varlist = c("aug_05", "aug_06", "aug_07", 
                                  "aug_08", "aug_09"),
                      timepts = c(0, 1, 2, 3, 4),
                      regtype = "lin",
                      int_bins = 3,
                      lin_bins = 3)
```

The OLStraj function provides several types of plots that facilitate the
examination of your data:

### Group Plots:

``` r
robins_traj$group_plots
#> $simple_joined
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

    #> 
    #> $ols
    #> `geom_smooth()` using formula = 'y ~ x'

<img src="man/figures/README-unnamed-chunk-6-2.png" width="100%" />

Group plots depict the simple-joined trajectories and the estimated
trajectories for each group in your data, allowing for a quick
comparison of trends across different groups.

### Individual OLS Trajectories:

``` r
robins_traj$individual_plots
#> $`ols Knobs_Flat`
#> `geom_smooth()` using formula = 'y ~ x'
```

<img src="man/figures/README-unnamed-chunk-7-1.png" width="100%" />

    #> 
    #> $`ols Walker_Creek`
    #> `geom_smooth()` using formula = 'y ~ x'

<img src="man/figures/README-unnamed-chunk-7-2.png" width="100%" />

Individual plots represent each subject’s unique trajectory over time,
providing insights into individual patterns of change.

### Histograms for intercepts and slopes:

``` r
robins_traj$histogram_plots
#> $intercepts
```

<img src="man/figures/README-unnamed-chunk-8-1.png" width="100%" />

    #> 
    #> $slopes

<img src="man/figures/README-unnamed-chunk-8-2.png" width="100%" />

These histograms reveal the distribution of intercepts and slopes across
subjects. They are useful for identifying skewness or other interesting
distributional properties in the data.

### Box plots of intercepts and slopes:

``` r
robins_traj$box_plot
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />

Box plots provide a summary of the distribution of intercepts and slopes
across subjects. They allow you to visualize the spread and skewness of
the data. Any outliers are clearly highlighted in these plots. In the
OLStraj box plots, the mean value is represented by a blue dot.
Additionally, each outlier is labeled with the corresponding subject
identifier, providing a straightforward way to identify subjects whose
data deviate significantly from the rest of the group.


NEWS.md
# OLStrajr 0.1.0

# OLStrajr 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(OLStrajr)

test_check("OLStrajr")


tests/testthat/test-cbc_lm.R
test_that("Test that cbc_lm returns an error for invalid input", {

  # Define some example data
  df <- data.frame(ids = rep(1:5, 5),
                   vals = stats::rnorm(25),
                   outs = stats::rnorm(25, 10, 25))

  # Test that error is returned when 'data' is not a data.frame
  expect_error(cbc_lm(data = c(1, 2, 3), formula = outs ~ vals, .case = "ids"), "'data' must be a data frame.")

  # Test that error is returned when 'formula' is not coercible to a formula
  expect_error(cbc_lm(data = df, formula = "not a formula", .case = "ids"), "'formula' must be a valid formula or a string coercible to a formula.")

  # Test that error is returned when '.case' is not a column name in 'data'
  expect_error(cbc_lm(data = df, formula = outs ~ vals, .case = "nonexistent_column"), "'.case' must be a valid column name in 'data'.")

  # Test that error is returned when 'n_bootstrap' is not a positive integer
  expect_error(cbc_lm(data = df, formula = outs ~ vals, .case = "ids", n_bootstrap = -5), "'n_bootstrap' must be a positive integer.")

  # Test that error is returned when 'lm_options', 'boot_options', or 'boot.ci_options' is not a list
  expect_error(cbc_lm(data = df, formula = outs ~ vals, .case = "ids", lm_options = c(1, 2, 3)), "'lm_options', 'boot_options', and 'boot.ci_options' must be lists.")

  # Test that error is returned when 'na.rm' is not a logical value
  expect_error(cbc_lm(data = df, formula = outs ~ vals, .case = "ids", na.rm = "not a logical value"), "'na.rm' must be a logical value.")
})

test_that("Test that cbc_lm runs correctly with valid input", {

  # Define some example data
  df <- data.frame(ids = rep(1:5, 5),
                   vals = stats::rnorm(25),
                   outs = stats::rnorm(25, 10, 25))

  # Test that the function runs without error
  expect_error(cbc_lm(data = df, formula = outs ~ vals, .case = "ids"), NA)

  # Test that the function returns an object of class "cbc_lm"
  expect_s3_class(cbc_lm(data = df, formula = outs ~ vals, .case = "ids"), "cbc_lm")

  # More specific tests can be added based on the expected outputs of the function
})


test_that("Test that cbc_lm gives expected output", {

  # Define some example data
  df <- data.frame(ids = rep(1:5, 5),
                   vals = stats::rnorm(25),
                   outs = stats::rnorm(25, 10, 25))


  res <- cbc_lm(data = df, formula = outs ~ vals, .case = "ids")

  expect_equal(length(res), 2)


  expect_equal(length(res$models), 5)

  expect_equal(length(summary(res)), 6)

  s_res <- summary(res)

  expect_null(s_res$bm_coef)

})


tests/testthat/test-OLStraj.R
test_that("OLStraj returns correct output for sample dataframe", {
  # Create a sample data frame
  df <- data.frame(id = c(1, 2, 3),
                   anti1 = c(1,3, 6),
                   anti2 = c(3, 5, 7),
                   anti3 = c(3, 4, 8),
                   anti4 = c(5, 6, 7))

  # Run the function
  result <- OLStraj(data = df,
                    varlist = c("anti1", "anti2", "anti3", "anti4"),
                    regtype = "lin",
                    int_bins = 5,
                    lin_bins = 5,
                    quad_bins = 5)

  # Assert that the output contains the expected elements
  expect_type(result, "list")
  expect_true("out_data" %in% names(result))
  expect_true("group_plots" %in% names(result))
  expect_true("individual_plots" %in% names(result))
})

test_that("OLStraj throws an error if data is not a dataframe", {
  # Assert that the function throws an error when the data argument is not a data frame
  expect_error(OLStraj(data = c(1, 2, 3)), "data must be a data frame.")
})

test_that("OLStraj throws an error if idvarname does not exist in data", {
  df <- data.frame(time = c(1, 2, 3), score = c(4, 5, 6))

  # Assert that the function throws an error when the idvarname argument does not exist in data
  expect_error(OLStraj(data = df), "idvarname must be a column name in the provided data frame.")
})

test_that("OLStraj throws an error if elements of varlist do not exist in data", {
  df <- data.frame(id = c(1, 2, 3), time = c(1, 2, 3), score = c(4, 5, 6))

  # Assert that the function throws an error when elements of varlist do not exist in data
  expect_error(OLStraj(data = df, varlist = c("nonexistent1", "nonexistent2")),
               "All elements of varlist must be column names in the provided data frame.")
})

test_that("OLStraj throws an error if inclmiss is not logical", {
  # Create a sample data frame
  df <- data.frame(id = c(1, 2, 3),
                   anti1 = c(1,3, 6),
                   anti2 = c(3, 5, 7),
                   anti3 = c(3, 4, 8),
                   anti4 = c(5, 6, 7))

  # Assert that the function throws an error when inclmiss is not logical
  expect_error(OLStraj(data = df, inclmiss = "not logical"), "inclmiss must be of type logical.")
})

# test_that("multiplication works", {
#   df <- data.frame(id = c(1,2,3,4,5),
#                    var1 = c(3,7,4,5,8),
#                    var2 = c(7,3,9,4,7),
#                    var3 = c(8,5,3,9,7),
#                    var4 = c(1,5,3,9,30))
#
#   olstraj_out <- OLStraj(data = df,
#                          varlist = c("var1", "var2", "var3", "var4"),
#                          regtype = "quad",
#                          int_bins = 5,
#                          lin_bins = 5,
#                          quad_bins = 5)
#
#   # Debugging parameters:
#   data = df
#   varlist = c("var1", "var2", "var3", "var4")
#   regtype = "both"
#   idvarname = "id"
#   predvarname = "time"
#   outvarname = "score"
#   timepts = c(0, 1, 2, 3)
#   inclmiss = "n"
#   level = "both"
#   numplot = NULL
#   hist = "y"
#   box = "y"
#   outds = TRUE
#
#   expect_equal(2 * 2, 4)
# })


R/data.R
#'  Ratio of robin males to females in Walker Creek and Knobs Flat, Eglinton Valley
#'
#'  Data from Table 1 of "Birds: incomplete counts—five-minute bird counts Version 1.0"
#'
#'
#' @format ## `robins`
#' A data frame with 2 observations and 6 variables:
#' \describe{
#'   \item{site}{Site name}
#'   \item{aug_05, aug_06, aug_07, aug_08, aug_09}{ratio of male to female robins}
#' }
#' @source <https://www.doc.govt.nz/documents/science-and-technical/inventory-monitoring/im-toolbox-birds-incomplete-five-min-counts.pdf>
"robins"


#'  Rat Weight Data from HLM manual
#'
#'  Data obtained from Rogosa & Saner (1995) which describes the data as:
#'  "Example 1. The rat weight data are taken from the HLM manual (Bryk et al., 1989).
#'  The rat data consist of 10 individuals, with weight measurements (Y) at five
#'  occasions (Weeks 0, 1,2, 3, 4) and a background measure, the mother's weight (Z)."
#'
#' @format ## `rats`
#' A data frame with 10 observations and 7 variables:
#' \describe{
#'   \item{Rat}{Rat identifier}
#'   \item{t0, t1, t2, t3, t4}{Week of weight measure}
#'   \item{Z}{Mother's weight}
#' }
#'
#' @references
#' Bryk, A. S., Raudenbush, S. W., Seltzer, M., & Congdon, R. T. (1989).
#' An introduction to HLM: Computer program and user's guide. Chicago: University of Chicago.
#' \doi{10.1201/9780429246593}
#'
#' Rogosa, D., & Saner, H. (1995). Longitudinal Data Analysis Examples with Random
#' Coefficient Models. Journal of Educational and Behavioral Statistics, 20(2), 149-170.
#' \doi{10.3102/10769986020002149}
#'
#' @source <https://www.taylorfrancis.com/books/mono/10.1201/9780429246593/introduction-bootstrap-bradley-efron-tibshirani>
"rats"


R/OLStraj.R
#' OLStraj
#'
#' Implements the OLS trajectory analysis method as detailed in Carrig et al (2004).
#' The method uses case-by-case ordinary least squares (OLS) regression to estimate
#' individual growth curves over time. The function provides options for
#' group-level and individual-level plots and accommodates linear and quadratic
#' models.
#'
#' @param data A data frame
#' @param idvarname A quoted variable name identifying the column in data which
#' serves as the case identifier
#' @param predvarname A quoted predictor variable label.
#' @param outvarname A quoted outcome variable label.
#' @param varlist A vector of quoted variable names found in data
#' @param timepts A vector specifying how time points should be coded
#' @param inclmiss A logical specifying whether or not to use complete cases.
#' Set inclmiss to FALSE in order to filter data down to complete cases.
#' @param level Control which OLS trajectory plots to show.  If level is set to
#' "grp" then only group level plots will be shown, if level is set to "ind" then
#' only individual level plots will be shown, and if level is set to "both" then
#' both group and individual level plots will be shown.
#' @param regtype Set regtype to "quad" to include quadratic term in the cbc_lm
#' call or set regtype to "lin" to exclude the quadratic term.  Use regtype = "both"
#' to include the quadratic term in the cbc_lm call and to include both linear
#' and quadratic terms on the individual OLS-estimated trajectory plots.
#' @param numplot Specify an integer to subset the number of cases used in OLStraj
#' @param hist Set hist to  TRUE to include histograms or FALSE to exclude
#' @param int_bins Set the number of bins for the intercept term's histogram
#' @param lin_bins Set the number of bins for the linear term's histogram
#' @param quad_bins Set the number of bins for the quadratic term's histogram
#' @param box Set box to TRUE to include boxplots or FALSE to exclude
#' @param outds Set outds to TRUE to include the output as a data frame.  Output
#' will contain original data used in the OLStraj algorithm with the parameter
#' estimates obtained from cbc_lm
#' @param ... Pass additional arguments to cbc_lm
#'
#' @return A list containing an output data frame (if outds is set to TRUE), the
#' selected plots, and the case-by-case regression model object.
#' @export
#'
#' @references
#' Carrig, M.M., Wirth, R.J., & Curran, P.J. (2004). A SAS Macro for Estimating
#' and Visualizing Individual Growth Curves.  Structural Equation Modeling: A
#' Multidisciplinary Journal, 11(1), 132-149.
#' \doi{10.1207/S15328007SEM1101_9}
#'
#' @examples
#'   df <- data.frame(id = c(1,2,3,4,5),
#'                    var1 = c(3,7,4,5,8),
#'                    var2 = c(7,3,9,4,7),
#'                    var3 = c(8,5,3,9,7),
#'                    var4 = c(1,5,3,9,30))
#'
#'   olstraj_out <- OLStraj(data = df,
#'                          varlist = c("var1", "var2", "var3", "var4"),
#'                          regtype = "quad",
#'                          int_bins = 5,
#'                          lin_bins = 5,
#'                          quad_bins = 5)
OLStraj <- function(data, idvarname = "id", predvarname = "time",
                    outvarname = "score",
                    varlist = c("anti1", "anti2", "anti3", "anti4"),
                    timepts = c(0, 1, 2, 3), inclmiss = FALSE, level = "both", regtype = "lin",
                    numplot = NULL, hist = TRUE, int_bins = 30, lin_bins = 30,
                    quad_bins = 30, box = TRUE, outds = TRUE, ...) {

  # Check if data is a data.frame
  if (!is.data.frame(data)) {
    stop("data must be a data frame.")
  }

  # Check if idvarname, predvarname and outvarname exist in data
  if (!(idvarname %in% names(data))) {
    stop("idvarname must be a column name in the provided data frame.")
  }

  # Check if varlist elements are column names in data
  if (!all(varlist %in% names(data))) {
    stop("All elements of varlist must be column names in the provided data frame.")
  }

  # Check if length of timepts matches length of varlist
  if(length(timepts) != length(varlist)){
    stop("ERROR: NUMBER OF TIME POINTS DOES NOT EQUAL NUMBER OF REPEATED MEASURES")
  }

  # Check if inclmiss is logical
  if (!is.logical(inclmiss)) {
    stop("inclmiss must be of type logical.")
  }

  # Check if level is one of "both", "grp", or "ind"
  if (!level %in% c("both", "grp", "ind")) {
    stop("level must be either 'both', 'grp', or 'ind'.")
  }

  # Check if regtype is either "lin" or "quad"
  if (!regtype %in% c("lin", "quad", "both")) {
    stop("regtype must be either 'lin', 'quad', or 'both'.")
  }

  # Check if numplot is NULL or an integer
  if (!is.null(numplot) & (!is.numeric(tryCatch(numplot, error = function(e) NULL) |
                                       (tryCatch(numplot != round(numplot), error = function(e) NULL))))) {
    stop("numplot must be either NULL or an integer.")
  }

  # Check if hist is logical
  if (!is.logical(hist)) {
    stop("hist must be of type logical.")
  }

  # Check if box is logical
  if (!is.logical(box)) {
    stop("box must be of type logical.")
  }

  # Check if outds is logical
  if (!is.logical(outds)) {
    stop("outds must be of type logical.")
  }

  # Check if int_bins, lin_bins, quad_bins are positive integers
  if (!is.numeric(int_bins) | (int_bins != round(int_bins)) | (int_bins <= 0) |
      !is.numeric(lin_bins) | (lin_bins != round(lin_bins)) | (lin_bins <= 0) |
      !is.numeric(quad_bins) | (quad_bins != round(quad_bins)) | (quad_bins <= 0)) {
    stop("int_bins, lin_bins, quad_bins must be positive integers.")
  }

  # Create subsample
  if (!is.null(numplot)) {
    data <- data[1:numplot, ]
  }

  # Save a copy of data in original format
  data_orig <- data

  # Listwise deletion
  if (inclmiss == FALSE) {
    data <- data[stats::complete.cases(data), ]
  }

  # OLS case-by-case regressions

  #Lengthen data frame
  data <- tidyr::pivot_longer(data, cols = tidyselect::all_of(varlist),
                              names_to = predvarname,
                              values_to = outvarname)

  data[[predvarname]] <- timepts[match(data[[predvarname]], varlist)]

  # Get quadratic term
  if (regtype != "lin"){
    data[[paste0(predvarname, "_sq")]] <- data[[predvarname]]^2
  }

  if (regtype == "lin"){
    cbc_form <- stats::as.formula(paste(outvarname, "~", predvarname))
  } else {
    cbc_form <- stats::as.formula(paste(outvarname, "~", predvarname, "+",
                                        paste0(predvarname, "_sq")))
  }

  models <- cbc_lm(data = data, formula = cbc_form, .case = idvarname, ...)

  ols_dat <- lapply(seq_along(models$models), function(i){
    # Define model
    model <- models$models[[i]]
    id_val <- names(models$models)[[i]]

    # Add the estimated values to the data frame
    if (regtype == "lin"){
      stats::setNames(data.frame(id_val,
                                 stats::coef(model)[1],
                                 stats::coef(model)[2],
                                 summary(model)[[8]]),
                      c(eval(idvarname), "intercept", "linear", "rsquared"))
    } else {
      stats::setNames(data.frame(id_val,
                                 stats::coef(model)[1],
                                 stats::coef(model)[2],
                                 stats::coef(model)[3],
                                 summary(model)[[8]]),
                      c(eval(idvarname), "intercept", "linear", "quad", "rsquared"))
    }
  })

  estimated_values <- do.call(rbind, ols_dat)

  # Write output to a data frame
  if (outds == TRUE) {
    out_data <- merge(data_orig, estimated_values, by = eval(idvarname))
  }

  # Plotting section
  group_plots <- list()
  individual_plots <- list()
  histogram_plots <- list()

  # Prepare data for plotting
  data <- merge(data, estimated_values, by = eval(idvarname))

  if (level == "both" | level == "grp") {
    # Group-level plots

    # Simple-joined (noninterpolated) trajectories
    group_plots[["simple_joined"]] <- ggplot2::ggplot(data,
                                                      ggplot2::aes(x = .data[[predvarname]],
                                                                   y = .data[[outvarname]],
                                                                   group = .data[[idvarname]])) +
      ggplot2::geom_line() +
      ggplot2::ggtitle("Simple-Joined Trajectories")

    # OLS trajectories
    group_plots[["ols"]] <- ggplot2::ggplot(data, ggplot2::aes(x = .data[[predvarname]], y = .data[[outvarname]],
                                                               group = .data[[idvarname]])) +
      ggplot2::ggtitle("OLS Trajectories")

    if (regtype == "lin"){
      group_plots[["ols"]] <- group_plots[["ols"]] +
        ggplot2::geom_smooth(se = FALSE, method = lm)

    } else {
      group_plots[["ols"]] <- group_plots[["ols"]] +
        ggplot2::stat_smooth(method = "lm", formula = y ~ x + I(x^2),
                             se = FALSE)
    }

  }

  if (level == "both" | level == "ind") {
    # Individual-level plots
    for (id in unique(data[[idvarname]])) {
      ind_data <- data[data[[idvarname]] == id, ]

      # OLS trajectories
      individual_plots[[paste("ols", id)]] <- ggplot2::ggplot(ind_data,
                                                              ggplot2::aes(x = .data[[predvarname]],
                                                                           y = .data[[outvarname]])) +
        ggplot2::ggtitle(paste("OLS Trajectory for", idvarname, ": ", id))

      if (regtype == "lin"){
        individual_plots[[paste("ols", id)]] <- individual_plots[[paste("ols", id)]] +
          ggplot2::geom_point() +
          ggplot2::geom_smooth(se = FALSE, method = lm)


      } else if (regtype == "quad"){
        individual_plots[[paste("ols", id)]] <- individual_plots[[paste("ols", id)]] +
          ggplot2::geom_point() +
          ggplot2::geom_smooth(method = "lm", formula = y ~ x + I(x^2), se = FALSE)

      } else {
        individual_plots[[paste("ols", id)]] <- individual_plots[[paste("ols", id)]] +
          ggplot2::geom_point() +
          ggplot2::geom_smooth(se = FALSE, method = lm) +
          ggplot2::geom_smooth(method = "lm", formula = y ~ x + I(x^2), se = FALSE,
                               linetype = "dashed")
      }
    }
  }

  if (hist == TRUE) {
    # Histogram
    intercepts <- ggplot2::ggplot(data, ggplot2::aes(x = intercept)) +
      ggplot2::geom_histogram(bins = int_bins) +
      ggplot2::ggtitle("Histogram of OLS Estimated Intercepts")

    slopes <- ggplot2::ggplot(data, ggplot2::aes(x = linear)) +
      ggplot2::geom_histogram(bins = lin_bins) +
      ggplot2::ggtitle("Histogram of OLS Estimated Slopes")

    if (regtype != "lin"){
      quads <- slopes <- ggplot2::ggplot(data, ggplot2::aes(x = quad)) +
        ggplot2::geom_histogram(bins = quad_bins) +
        ggplot2::ggtitle("Histogram of OLS Estimated Quadratic Terms")

      histogram_plots = list("intercepts" = intercepts,
                             "slopes" = slopes,
                             "quads" = quads)
    } else {
      histogram_plots = list("intercepts" = intercepts,
                             "slopes" = slopes)
    }
  }

  if (box == TRUE) {
    #Add mean to boxplot and label outliers:
    if (regtype == "lin"){
      data_box <- data |> tidyr::pivot_longer(cols = c(intercept, linear),
                                              names_to = "param",
                                              values_to = "est")
    } else {
      data_box <- data |> tidyr::pivot_longer(cols = c(intercept, linear, quad),
                                              names_to = "param",
                                              values_to = "est")
    }

    # calculate the upper and lower bounds for outliers
    boxplot_stats <- tapply(data_box$est, data_box$param, grDevices::boxplot.stats)
    outlier_vals <- lapply(boxplot_stats, `[[`, "out")

    # Initialize outlier column
    data_box$outlier <- NA

    # Loop over the parameters and identify outliers
    for(i in unique(data_box$param)) {
      pick_outliers <- data_box$param == i & data_box$est %in% outlier_vals[[i]] &
        is.na(data_box$outlier)

      data_box$outlier[pick_outliers] <- data_box[[idvarname]][pick_outliers]
    }

    # Convert outlier column to character and change missing to ""
    data_box$outlier <- as.character(data_box$outlier)
    data_box$outlier <- ifelse(is.na(data_box$outlier), "", data_box$outlier)

    #Create box plot with mean plotted and outliers labeled by idvarname
    data_box <- data_box |>
      ggplot2::ggplot(ggplot2::aes(x = param, y = est)) +
      ggplot2::geom_boxplot() +
      ggplot2::stat_summary(fun = mean, geom = "point", shape = 20, size = 3, color = "blue") + # Add point for mean
      ggplot2::geom_text(ggplot2::aes(label = outlier), hjust = -0.1) +
      ggplot2::ylab("OLS Estimates") +
      ggplot2::xlab("") +
      ggplot2::ggtitle("Box Plots of OLS Estimates")
  }


  # Return a list containing estimated values and the plots
  result <- list(out_data = out_data,
                 group_plots = group_plots,
                 individual_plots = individual_plots,
                 histogram_plots = histogram_plots,
                 box_plot = data_box,
                 models = models)

  return(result)
}
